storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3088,6 +3119,690 @@ std::ostream &fseg_header::to_stream(std::ostream &out) const
+    uint32_t cur_extent= (cur_page_no / FSP_EXTENT_SIZE) * FSP_EXTENT_SIZE;
+    uint32_t state= space_defrag->get_state(cur_extent);
+
+    if (state == XDES_FREE)
+    {
+fetch_next_page:
+      mtr.commit();
+
+      if (next_page_no == FIL_NULL)
+        break;
+      cur_page_no= next_page_no;
+
+      mtr.start();
+      mtr_x_lock_index(m_index, &mtr);
+      mtr.x_lock_space(fil_system.sys_space);
+      block= buf_page_get_gen(page_id_t{0, cur_page_no}, 0, RW_X_LATCH,
+                            nullptr, BUF_GET_POSSIBLY_FREED, &mtr,
+                            &err);
+      continue;
+    }
+
+    uint32_t new_extent= space_defrag->get_new_extent(cur_extent);
+
+    if (new_extent == cur_extent)
+    {
+next_page:
+      /* Read all child pages exist in parent block */
+      if (level)
+        store_child_pages(block);
+      goto fetch_next_page;
+    }
+
+    /* Need to replace the current extent with new extent */
+    buf_block_t *root= btr_root_block_get(m_index, RW_SX_LATCH, &mtr,
+                                          &err);
+    if (UNIV_UNLIKELY(!root))
+      goto err_exit;
+
+    buf_block_t *new_block= nullptr;
+    fseg_header_t *seg_header=
+      root->page.frame + (level
+                          ? PAGE_HEADER + PAGE_BTR_SEG_TOP
+                          : PAGE_HEADER + PAGE_BTR_SEG_LEAF);
+
+    buf_block_t *iblock;
+    fseg_inode_t *inode= fseg_inode_try_get(seg_header, 0, 0, &mtr,
+                                            &iblock, &err);
+    if (!inode)
+      goto err_exit;
+
+    switch(state)
+    {
+    case XDES_FSEG:
+      new_block=
+	space_defrag->alloc_page_from_new_seg(seg_header, iblock,
+                                              inode, new_extent,
+                                              &mtr, &err);
+    break;
+    case XDES_FREE_FRAG:
+    case XDES_FULL_FRAG:
+      new_block= fsp_alloc_free_page<true>(space, new_extent,
+                                           &mtr, &mtr, &err);
+    break;
+    case XDES_FREE:
+      ut_error;
+    }
+
+    if (!new_block)
+      goto err_exit;
+
+    /** Copy the contents from block to new block */
+    mtr.memcpy(*new_block, new_block->page.frame,
+               block->page.frame, srv_page_size - FIL_PAGE_DATA_END);
+
+    uint32_t new_page_no= new_block->page.id().page_no();
+    mtr.write<4,mtr_t::FORCED>(*new_block,
+                               FIL_PAGE_OFFSET + new_block->page.frame,
+                               new_page_no);
+
+    uint32_t prev_page_no= btr_page_get_prev(page);
+
+    /* change the previous page */
+    if (prev_page_no != FIL_NULL)
+    {
+      buf_block_t *prev_block=
+        buf_page_get_gen(page_id_t{0, prev_page_no}, 0, RW_X_LATCH,
+                         nullptr, BUF_GET_POSSIBLY_FREED, &mtr,
+                         &err);
+      if (!prev_block)
+        goto err_exit;
+      mtr.write<4>(*prev_block,
+                   prev_block->page.frame + FIL_PAGE_NEXT,
+                   new_page_no);
```

Comment:
All the error handling needs to be rewritten in such a way that we first check all prerequisites, and only then modify any metadata or data. For example here, if we fail to access `next_block` later, we’d already have modified `prev_block` and the `if (!next_block) goto err_exit` would leave the database in a corrupted state.